pkg/generator/secret_template_reconciler.go

```diff
@@ -0,0 +1,272 @@
+	"k8s.io/apimachinery/pkg/types"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	// Matches default sync in kapp-controller
+	// See: https://github.com/vmware-tanzu/carvel-kapp-controller/blob/develop/pkg/app/reconcile_timer.go
+	syncPeriod = 30 * time.Second
+)
+
+// ClientLoader allows Kubernetes Clients to be loaded from a Service Account.
+type ClientLoader interface {
+	Client(ctx context.Context, saName, saNamespace string) (client.Client, error)
+}
+
+// SecretTemplateReconciler watches for SecretTemplate Resources and generates a new secret from a set of input resources.
+type SecretTemplateReconciler struct {
+	client   client.Client
+	saLoader ClientLoader
+	log      logr.Logger
+}
+
+var _ reconcile.Reconciler = &SecretTemplateReconciler{}
+
+// NewSecretTemplateReconciler create a new SecretTemplate Reconciler
+func NewSecretTemplateReconciler(client client.Client, loader ClientLoader, log logr.Logger) *SecretTemplateReconciler {
+	return &SecretTemplateReconciler{client, loader, log}
+}
+
+// AttachWatches adds starts watches this reconciler requires.
+func (r *SecretTemplateReconciler) AttachWatches(controller controller.Controller) error {
+	//Watch for changes to created Secrets
+	if err := controller.Watch(&source.Kind{Type: &corev1.Secret{}}, &handler.EnqueueRequestForOwner{OwnerType: &sg2v1alpha1.SecretTemplate{}}); err != nil {
+		return err
+	}
+	return controller.Watch(&source.Kind{Type: &sg2v1alpha1.SecretTemplate{}}, &handler.EnqueueRequestForObject{})
+}
+
+// Reconcile is the entrypoint for incoming requests from k8s
+func (r *SecretTemplateReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("request", request)
+
+	secretTemplate := sg2v1alpha1.SecretTemplate{}
+	if err := r.client.Get(ctx, types.NamespacedName{Namespace: request.Namespace, Name: request.Name}, &secretTemplate); err != nil {
+		if errors.IsNotFound(err) {
+			log.Info("Not found")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	if secretTemplate.DeletionTimestamp != nil {
+		return reconcile.Result{}, nil
+	}
+
+	status := &reconciler.Status{
+		S:          secretTemplate.Status.GenericStatus,
+		UpdateFunc: func(st sgv1alpha1.GenericStatus) { secretTemplate.Status.GenericStatus = st },
+	}
+
+	status.SetReconciling(secretTemplate.ObjectMeta)
+	defer r.updateStatus(ctx, &secretTemplate)
+
+	res, err := r.reconcile(ctx, &secretTemplate)
+	//TODO is this overly defensive?
+	if err != nil {
+		secretTemplate.Status.Secret.Name = ""
+		if deleteErr := r.deleteChildSecret(ctx, &secretTemplate); deleteErr != nil {
+			return status.WithReconcileCompleted(res, deleteErr)
+		}
+	}
```

Comment:
yeah, im also on the fence about this...

i fear deletion behaviour will cause instability and outages. (e.g. im changing my SecretTemplate to ref secret2 instead of secret1, while at the same time deleting secret1. if it happens that secret1 is deleted first, my result secret is gone which may be is being consumed by some app that now has an outage. when secret2 is brought back in, result secret is "fixed"; or if you have screwed up your templating your secret gets deleted and again your apps potentially incur outage).

i think generally i would side with system should not change created resources if it cannot construct them successfully (in our case due to templating errors).

> We thought it was risky to keep the generated Secret around with potentially sensitive data that the user had thought they had deleted from the cluster.

once the data touches the cluster all bets are off in my opinion. removing it will not "undo the damage" -- if something is truly sensitive and it's leaked, it should be invalidated with a system that cares about that credential.

is there some precedent in k8s core or other projects?

Reply:
@cppforlife no prior art that I'm aware of. Happy to go with option 3 for this PR. We can always improve this in the future.